### ISSUE TITLE
Changes to how/when the `loss` is calculated by `forward` and `validate` in `ComposerTransformers` 

### ISSUE THREAD

**Issue by jzf2101** [mosaicml/composer#993]
## 🚀 Feature Request

Additional consideration into how/when `loss` is calculated in `forward` and `validate`.  HF does not give access to the loss itself and the loss is transformer variant specific.

## Motivation

1. Inconsistent inputs/outputs and calculation method for `validate` in `GPT2Model` and `BertModel`:
a. `BERTModel`'s `validate` method does not calculate the loss because it pops the `labels` and instead only returns the tensors `logits` and tensor of popped `labels`. 
b. `GPT2Model`'s `validate` method does *not* pop the labels and returns a  `CausalLMOutputWithCrossAttentions` object with the losses and labels

2. HF transformers do not give us access to the `loss` as a method, only the reduced `CrossEntropy` loss scalar after `forward` is run.  However, for methods like `selective_backprop` the loss is expected to be callable from the model. Moreover, the loss values are expected to be unreduced.3. 
a. Note that the loss calculation is also transformer model class specific, and will have to be reimplemented from scratch based on the `logits` returned from the forward.

## [Optional] Implementation

For Issue 1: 

Option 1a: remove `labels` from `batch` in `ComposerTransformer`'s `validate` method so this is consistent across *all* transformers.

Option 1b: change `validate` in `GPT2Model` to match `validate` in BertModel.

For Issue 2:

Option 2a: Give `GPT2Model` and `BERTModel` a method to return the unreduced loss

Option 2b: Let `selective_backprop` specifically handle recalculating the unreduced loss since its the only algorithm that I can think of that needs it now

## Additional Context

These concerns have been previously discussed with @growlix and @moinnadeem

**Comment by moinnadeem** [comment]
+1 on this -- I've been wondering if we should move loss from HF's codebase to ours, but didn't have a motivating reason to, so it felt like premature optimization. I think selective backprop is now that reason.

I view the loss being Transformer-specific as a feature, since we don't need to think about the complexity there and HF handles it for us.

Part of this might intersect with @Landanjs's proposal for adding loss function selection into Composer. Are we proceeding with that>

**Comment by growlix** [comment]
Seems to me like there's an expedient solution to 1: Have `selective_backprop` type check the output of `validate()` and extract the data accordingly. Though I could be mistaken about exactly how `validate()` works!

**Comment by mvpatel2000** [comment]
Closing, we won't impose this on all models
